src/signals-unix.c

```diff
@@ -131,72 +140,80 @@ static void allocate_segv_handler(void)
+    jl_tls_states_t *ptls = jl_all_task_states[tid].ptls;
+    jl_atomic_store_release(&ptls->signal_request, 1);
     pthread_cond_broadcast(&exit_signal_cond);
     pthread_cond_wait(&signal_caught_cond, &in_signal_lock); // wait for thread to acknowledge
-    assert(waiting_for == 0);
+    assert(jl_atomic_load_acquire(&ptls->signal_request) == 0);
     pthread_mutex_unlock(&in_signal_lock);
 }
 
-
-static inline void wait_barrier(void)
+// Throw jl_interrupt_exception if the master thread is in a signal async region
+// or if SIGINT happens too often.
+static void jl_try_deliver_sigint(void)
 {
-    if (waiting_for < 0) {
-        if (jl_atomic_fetch_add(&waiting_for, 1) == -1) {
-            pthread_cond_broadcast(&signal_caught_cond);
-        }
-    }
-    else {
-        pthread_cond_broadcast(&signal_caught_cond);
-        waiting_for = 0;
-    }
+    jl_tls_states_t *ptls = jl_all_task_states[0].ptls;
+    jl_safepoint_enable_sigint();
+    jl_wake_libuv();
+    jl_atomic_store_release(&ptls->signal_request, 2);
+    // This also makes sure `sleep` is aborted.
+    pthread_kill(jl_all_task_states[0].system_id, SIGUSR2);
 }
+
+// request:
+// 0: nothing
+// 1: get state
+// 3: throw sigint if `!defer_signal && io_wait` or if force throw threshold
+//    is reached
 void usr2_handler(int sig, siginfo_t *info, void *ctx)
 {
     ucontext_t *context = (ucontext_t*)ctx;
-    if ((remote_sig > 0 && waiting_for < 0) || waiting_for == ti_tid) {
-        int realsig = remote_sig;
+    jl_tls_states_t *ptls = jl_get_ptls_states();
+    sig_atomic_t request = jl_atomic_exchange(&ptls->signal_request, 0);
+    if (request == 1) {
 #ifdef __APPLE__
         signal_context = (unw_context_t*)&context->uc_mcontext->__ss;
 #else
         signal_context = (unw_context_t*)context;
 #endif
 
         pthread_mutex_lock(&in_signal_lock);
-        wait_barrier();
+        pthread_cond_broadcast(&signal_caught_cond);
         pthread_cond_wait(&exit_signal_cond, &in_signal_lock);
-        wait_barrier();
+        request = jl_atomic_exchange(&ptls->signal_request, 0);
+        assert(request == 1);
+        if (request)
+            pthread_cond_broadcast(&signal_caught_cond);
         pthread_mutex_unlock(&in_signal_lock);
-
-        if (ti_tid == 0 && realsig == SIGINT) {
-            if (jl_defer_signal) {
-                jl_signal_pending = realsig;
-            }
-            else {
-                jl_signal_pending = 0;
-                jl_unblock_signal(sig);
-                jl_throw(jl_interrupt_exception);
-            }
+    }
+    else if (request == 2) {
+        jl_unblock_signal(sig);
+        int force = jl_check_force_sigint();
+        if (force || (!ptls->defer_signal && ptls->io_wait)) {
+            jl_safepoint_consume_sigint();
+            if (force)
+                jl_safe_printf("WARNING: Force throwing a SIGINT\n");
+            // Force a throw
+            jl_clear_force_sigint();
+            jl_throw(jl_interrupt_exception);
```

Comment:
maybe add a TODO to replace this with `jl_throw_in_ctx` (after writing said function)